node/src/components/chain_synchronizer/operations.rs

```diff
@@ -684,23 +738,24 @@ async fn sync_deploys_and_transfers_and_state(
 ///
 /// Returns the block header with our current version and the last trusted key block information for
 /// validation.
-async fn archival_sync(
-    effect_builder: EffectBuilder<JoinerEvent>,
-    trusted_block_header: BlockHeader,
-    chainspec: &Chainspec,
-    node_config: NodeConfig,
-) -> Result<(KeyBlockInfo, BlockHeader), Error> {
+async fn archival_sync(ctx: &ChainSyncContext<'_>) -> Result<(KeyBlockInfo, BlockHeader), Error> {
     // Get the trusted block info. This will fail if we are trying to join with a trusted hash in
     // era 0.
+    info!("start - archival_sync - total");
     info!("start - fetch the latest key block before the trusted block - archival sync");
     let mut trusted_key_block_info =
-        get_trusted_key_block_info(effect_builder, chainspec, &trusted_block_header).await?;
+        get_trusted_key_block_info(*ctx.effect_builder, ctx.chainspec, ctx.trusted_block_header)
+            .await?;
     debug!("finish - fetch the latest key block before the trusted block - archival sync");
 
     info!("start - fetch the trusted block - archival sync");
     let trusted_block = *fetch_and_store_block_by_hash(
-        effect_builder,
-        trusted_block_header.hash(chainspec.protocol_config.verifiable_chunked_hash_activation),
+        *ctx.effect_builder,
+        ctx.trusted_block_header.hash(
+            ctx.chainspec
+                .protocol_config
+                .verifiable_chunked_hash_activation,
+        ),
```

Comment:
Actually, don't we have the trusted _hash_ in the node config? Then we wouldn't need to recompute it here.
Otherwise, maybe we should just add it to `ChainSyncContext`?

Reply:
Ah, you're right, we need to accept it from either config or storage. Happy to leave it as it is.

Reply:
Hmm, so we can actually be fetching the same block header twice:
1. in the constructor for `ChainSynchronizer` to read the trusted hash
2. in `run_chain_sync_task()` to read the trusted block header by trusted hash

We could possibly remove the `trusted_hash` parameter from the `start_syncing()` and `run_chain_sync_task()` functions and let the latter read it by itself (from config or from latest block).

Reply:
The `trusted_hash` is now part of `ChainSyncContext` and reused: ceea7dba7af6914a226021035a89916d7f299196